taichi/transforms/constant_fold.cpp

```diff
@@ -177,11 +177,32 @@ class ConstantFold : public BasicStmtVisitor {
       auto evaluated =
           Stmt::make<ConstStmt>(LaneAttribute<TypedConstant>(new_constant));
       stmt->replace_with(evaluated.get());
-      modifier.insert_before(stmt, VecStatement(std::move(evaluated)));
+      modifier.insert_before(stmt, std::move(evaluated));
       modifier.erase(stmt);
     }
   }
 
+  void visit(BitExtractStmt *stmt) override {
+    auto input = stmt->input->cast<ConstStmt>();
+    if (!input)
+      return;
+    if (stmt->width() != 1)
+      return;
+    int32 result;
+    if (is_signed(input->val[0].dt)) {
+      result = (input->val[0].val_int() >> stmt->bit_begin) &
+               ((1LL << (stmt->bit_end - stmt->bit_begin)) - 1);
+    } else {
+      result = (input->val[0].val_uint() >> stmt->bit_begin) &
+               ((1LL << (stmt->bit_end - stmt->bit_begin)) - 1);
+    }
+    auto result_stmt = Stmt::make<ConstStmt>(
+        LaneAttribute<TypedConstant>(TypedConstant(DataType::i32, result)));
```

Comment:
Given the part before already uses `uint` and `int` which covers all integral types, here maybe we should support all integral types as well? Just change `DataType::i32` into the input data type.

Reply:
Or we may need an assertion on `input_type != i8 && :input_type != i16`?

Reply:
> Given the part before already uses `uint` and `int` which covers all integral types, here maybe we should support all integral types as well? Just change `DataType::i32` into the input data type.

The `typecheck` pass set `ret_type = i32` for `OffsetAndExtractBitsStmt`. Shall we also change it to the input data type?

Reply:
> Or we may need an assertion on `input_type != DataType::i8 && input_type != DataType::i16`?

I think this is OK -- if the input is representable by `i8`, so is the output.